File: python/semantic_kernel/agents/runtime/core/agent_id.py
# Copyright (c) Microsoft. All rights reserved.

import sys
from typing import Protocol, runtime_checkable

from semantic_kernel.utils.feature_stage_decorator import experimental

if sys.version < "3.11":
    from typing_extensions import Self  # pragma: no cover
else:
    from typing import Self  # type: ignore # pragma: no cover

from semantic_kernel.agents.runtime.core.agent_type import AgentType
from semantic_kernel.agents.runtime.core.validation_utils import is_valid_agent_type


@experimental
@runtime_checkable
class AgentId(Protocol):
    """Defines the minimal interface an AgentId.

    It must fulfill a 'type' and a 'key' that identify the agent instance.
    """

    @property
    def type(self) -> str:
        """Defines the 'type' or category of the agent."""
        ...

    @property
    def key(self) -> str:
        """Defines the unique instance key within the agent type."""
        ...

    def __eq__(self, other: object) -> bool:
        """Equality check must differentiate between different IDs."""
        ...

    def __hash__(self) -> int:
        """Hash value needed to store AgentIds in sets/dicts."""
        ...

    def __str__(self) -> str:
        """String representation of the AgentId, e.g. 'type/key'."""
        ...


@experimental
class CoreAgentId(AgentId):
    """Core implementation of the AgentId protocol."""

    def __init__(self, type: str | AgentType, key: str) -> None:
        """Initialize the AgentId with the given type and key."""
        # If `type` is itself an AgentType, extract the string property.
        if isinstance(type, AgentType):
            type = type.type

        if not is_valid_agent_type(type):
            raise ValueError(
                rf"Invalid agent type: {type}. "
                r"Allowed values MUST match the regex: `^[\w\-\.]+\Z`"
            )

        self._type = type
        self._key = key

    @classmethod
    def from_str(cls, agent_id: str) -> Self:
        """Convert a string of the format ``type/key`` into a CoreAgentId."""
        items = agent_id.split("/", maxsplit=1)
        if len(items) != 2:
            raise ValueError(f"Invalid agent id: {agent_id}")
        t, k = items[0], items[1]
        return cls(t, k)

    @property
    def type(self) -> str:
        r"""The agent's 'type' (or category). Must match `^[\\w\\-\\.]+$`."""
        return self._type

    @property
    def key(self) -> str:
        """The agent's instance key, e.g. 'default' or a unique identifier."""
        return self._key

    def __eq__(self, value: object) -> bool:
        """Check if two AgentIds are equal by comparing 'type' and 'key'."""
        if not isinstance(value, AgentId):
            return False
        return (self.type == value.type) and (self.key == value.key)

    def __hash__(self) -> int:
        """Generate a hash so we can store AgentIds in sets/dicts."""
        return hash((self._type, self._key))

    def __str__(self) -> str:
        """Convert the AgentId to a user-friendly string."""
        return f"{self._type}/{self._key}"

    def __repr__(self) -> str:
        """Generate a detailed string representation."""
        return f'CoreAgentId(type="{self._type}", key="{self._key}")'

File: nodejs/semantic-kernel/agents/runtime/core/agent-id.ts
import { experimental } from '../../../utils/feature-stage-decorator'
import { AgentType } from './agent-type'
import { isValidAgentType } from './validation-utils'

/**
 * Defines the minimal interface an AgentId.
 *
 * It must fulfill a 'type' and a 'key' that identify the agent instance.
 *
 * Note: This interface is marked as 'experimental' and may change in the future.
 */
export interface AgentId {
  /**
   * Defines the 'type' or category of the agent.
   */
  readonly type: string

  /**
   * Defines the unique instance key within the agent type.
   */
  readonly key: string

  /**
   * Equality check must differentiate between different IDs.
   */
  equals(other: AgentId): boolean

  /**
   * String representation of the AgentId, e.g. 'type/key'.
   */
  toString(): string
}

/**
 * Core implementation of the AgentId interface.
 *
 * Note: This class is marked as 'experimental' and may change in the future.
 */
@experimental
export class CoreAgentId implements AgentId {
  private readonly _type: string
  private readonly _key: string

  constructor(type: string | AgentType, key: string) {
    // If `type` is itself an AgentType, extract the string property.
    if (typeof type === 'object' && 'type' in type) {
      type = type.type
    }

    if (!isValidAgentType(type)) {
      throw new Error(`Invalid agent type: ${type}. Allowed values MUST match the regex: ^[\\w\\-\\.]+$`)
    }

    this._type = type
    this._key = key
  }

  /**
   * Convert a string of the format `type/key` into a CoreAgentId.
   */
  static fromString(agentId: string): CoreAgentId {
    const items = agentId.split('/', 2)
    if (items.length !== 2) {
      throw new Error(`Invalid agent id: ${agentId}`)
    }
    const [type, key] = items
    return new CoreAgentId(type, key)
  }

  /**
   * The agent's 'type' (or category). Must match `^[\\w\\-\\.]+$`.
   */
  get type(): string {
    return this._type
  }

  /**
   * The agent's instance key, e.g. 'default' or a unique identifier.
   */
  get key(): string {
    return this._key
  }

  /**
   * Check if two AgentIds are equal by comparing 'type' and 'key'.
   */
  equals(other: AgentId): boolean {
    return this.type === other.type && this.key === other.key
  }

  /**
   * Convert the AgentId to a user-friendly string.
   */
  toString(): string {
    return `${this._type}/${this._key}`
  }
}
